### ISSUE TITLE
SearchBar props and types

### ISSUE THREAD

**Issue by kristojorg** [NYPL/nypl-design-system#220]
### Component in question:
- SearchBar
- Also Input

### Description of issue:
I would like to use the SearchBar in CPW, however there are a few changes that I think would be great to make before I did that:

1. Remove the mandatory IDs (opened another issue for this with all components)
2. rename `searchBarAriaLabel` and `searchBarAriaLabelledBy` to simply use the default `aria-label` and `aria-labelledby` names. This is a trend I think we should follow in general, using the native prop names where applicable. 
3. rename `searchValue` to simply `value`. Same idea as above
4. rename `placeHolder` to `placeholder`
5. rename the handlers from `searchChangeHandler` and `selectChangeHandler` to `onChange` and `onSelectChange`.
6. Refactor the error states into typescript. There are currently errors thrown for incorrect props, but those requirements can be expressed in typescript instead. 

Those are the more minor changes, but there is a larger rethinking of this component I would like to propose as well: I want to consider refactoring the `SearchBar` API to simply be a component composed from an `Input`, `Select` and `Button`. 

The `SearchBar` component doesn't introduce any logic per-se, it's mainly styling. I think it would be easier to let the consuming app compose one themself from smaller components. This way the consumer doesn't have to get to know the specifics of this API and how it passes props through to children. An example API composing a `SearchBar` from `Input`, `Button`, and `Select`:

```js
const SearchBar = () => {
  const [value, setValue] = React.useState("");
  const [selectValue, setSelectValue] = React.useState("keyword");
  const handleChange = (e) => {
    setValue(e.target.value);
  }
  const handleSelectChange = (e) => {
    setSelectValue(e.target.value)
  }
  const handleSearch = () => {
    api.search(value, selectValue);
  }
  
  return (
    <InputGroup aria-label="Search">
      <InputLeftAddon >
        <Select aria-label="Search Type" onChange={handleSelectChange}>
      </InputLeftAddon>
      <Input placeholder="Enter a search term..." onChange={handleChange} />
     <InputRightAddon >
       <Button aria-label="Submit search" onClick={handleSearch}>Search</Button>
     </InputRightAddon>
    </InputGroup>
  );
}
```

This is inspired by [Chakra UI](https://chakra-ui.com/input#left-and-right-addons), where the `InputLeftAddon` and `InputRightAddon` position their children correctly with the input so it appears similar to how the NYPL one does. 

Hopefully from this example it's more clear what I mean by how it simplifies the API for the consumer. The consuming dev doesn't have to learn `selectBlurHandler` and `dropdownOptions`, they just use the `Select` as they normally would, with an `onChange` and pass it to the `InputGroup` so it can be positioned correctly. 

None of this is stopping me from being able to use the component right now, but I think this would be a worthwhile change to make.

**Comment by hokei** [comment]
Lots of great stuff here -- `This is a trend I think we should follow in general, using the native prop names where applicable.` 
Yes, let's do this.   

I'm a bit concerned about 
```
  const handleChange = (e) => {
    setValue(e.target.value);
  }
  const handleSelectChange = (e) => {
    setSelectValue(e.target.value)
  }
  const handleSearch = () => {
    api.search(value, selectValue);
  }
```

The reason why we're passing all of these is because it's not clear what the consumer wants to do with them.  For example, with `handleSearch`, they may want to first log to analytics, and then search.  If we make them pass the api, then they won't be able to do this.  

Or, something like 
`<Button aria-label="Submit search" onClick={handleSearch}>Search</Button>` 

people are going to want to change the aria-label, and the wording on the Search button.

And I think we're going to get back into a situation where we have a whole bunch of props and people have to pass them in at the same time.

**Comment by hokei** [comment]
In contrast with the above, where I feel like putting the handlers inside DS is too specific, I feel like `<InputLeftAddon>` and `<InputRightAddon>` are going to be too general.  All our searchbars are going to have a dropdown on the left and a button on the right - if there's a searchbar that is commonly used that doesn't follow this, we're going to call it like, "WorkSearch" vs "SiteSearch" or something, since this design system is so specific to NYPL.

We don't really want people to be able to compose anything, we want them to be able to go, "I want a searchbar, and here are the searchbars that are available to me".  So in some ways, I think it's not that necessary to make them extensible, we just need to make the crazy amount of props manageable, and I think https://github.com/NYPL/nypl-design-system/issues/191 is my attempt at thinking through how we might do that.  I'm completely up for suggestions though - I wasn't able to find examples of other things that are trying to do what we are doing, and I definitely don't think this is the best way.

**Comment by kristojorg** [comment]
> I'm a bit concerned about
> 
> ```
>   const handleChange = (e) => {
>     setValue(e.target.value);
>   }
>   const handleSelectChange = (e) => {
>     setSelectValue(e.target.value)
>   }
>   const handleSearch = () => {
>     api.search(value, selectValue);
>   }
> ```

Ah sorry @hokei , this is what I meant the _consumer_ API could look like. So for example, this is what the code would look like in CPW to build our `SearchBar`. So technically there would no longer be a `SearchBar` within the DS. If you wanted to provide a default `SearchBar` though, you could build one like the above but yes let the consumer pass in the change handlers and labels. I was suggesting though that we basically get rid of `SearchBar` and just let the consumer piece it together like the above however they want to. The handlers there were just examples of what they might do.

**Comment by kristojorg** [comment]
Regarding not wanting it to be quite that flexible: I hear you. That is a valid point. I'll read through your suggestions in #191

**Comment by hokei** [comment]
`Refactor the error states into typescript. There are currently errors thrown for incorrect props, but those requirements can be expressed in typescript instead.` 

Yeah, we should get rid of those.  

There's a few things I haven't been able to think through: 
1) A lot of times people want to pass extra attributes (`button name` is an example I've had to deal with recently).  Is it a better situation to allow everything in https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button as a prop by defining them here and making them all optional, or should we have a catch-all `attributes` prop where people can pass it?  

2) Prop typos don't get caught from the consumer side.  I'm not sure if it's because SFR (the only current consumer) is a JS app instead of a TS one.  But it makes it really frustrating to develop on because stuff will not work and it's because you missed a letter somewhere.

**Comment by kristojorg** [comment]
Ah ok so **RE 1**:

Generally, when making a component that builds upon a base HTML element, I reuse the types for that element in the component I'm building, and possibly will extend them. For example, here is the `Select` component in CPW, which basically just applies styling right now:
```tsx
type SelectProps = React.ComponentProps<"select">;
const Select = React.forwardRef<HTMLSelectElement, SelectProps>(
  ({ children, ...props }, ref) => {
    return (
      <select ref={ref} sx={{ variant: "inputs.select" }} {...props}>
        {children}
      </select>
    );
  }
);
```
`React.ComponentProps<"select">` does some heavy lifting there by extracting the props that a native html `<select>` can have. You can do the same with components instead of elements, and extend them with your own props:
```ts
// in general in CPW, the button components take a variant, and extend some other component
type GenericButtonProps<T> = React.ComponentType<T extends React.ComponentType> 
  & { variant: "filled" | "outline" };
// a Button is like a normal <button> and it takes a normal 
// onClick and all the other HTML attributes
type ButtonProps = GenericButtonProps<"button">;
// a LinkButton looks and works like a Button, and takes variants but instead
// of <button> attributes like onClick, it takes props from React-Router Link, like "to"
type LinkButtonProps = GenericButtonProps<Link>;
```
Maybe a weird example, but you can see how you can use it to extract the props of another component or element so you don't have to define them all yourself.

Additionally, this pattern of keeping relatively close to the native HTML api is something I strive for when building UI components. That's part of why I prefer an API like
```jsx
<List>
  <List.ListItem aria-label="something">blah</List.ListItem>
</List>
```
over 
```jsx
<List items={myItems} /> // how would we pass an aria-label to a child?
```
It makes it easier to know what can be passed where and how. We all know how to use the native DOM, and unless it's a pretty complex component, most UI components can reuse the existing DOM api and just provide some extra functionality or styling.

**RE 2:**

That makes sense, and yea I think it's because it's a JS app. In that case maybe you keep the errors for now, though I imagine there is some kind of package to check props match typescript types at runtime in dev mode, to make it easier for you in the future. Maybe. Idk. If not someone should build that : )

**Comment by helenvholmes** [comment]
Just discussed this component with Bill for the Location Finder work. Sounds like the Reno team is going to request it have autocomplete functionality, so it might be best to bundle that work in here as well.

**Comment by helenvholmes** [comment]
Closed by #274.
